CSharpMath/Structures/Dictionary.cs

```diff
@@ -125,33 +93,46 @@ public class LaTeXCommandDictionary<TValue> : ProxyAdder<string, TValue>, IEnume
+    public Result<(TValue Result, int SplitIndex)> TryLookup(ReadOnlySpan<char> chars) {
+      if (chars.IsEmpty) throw new ArgumentException("There are no characters to read.", nameof(chars));
+      if (chars.StartsWithInvariant(@"\")) {
+        var splitIndex = SplitCommand(chars);
+        var lookup = chars.Slice(0, splitIndex);
+        while (splitIndex < chars.Length && char.IsWhiteSpace(chars[splitIndex]))
           splitIndex++;
         return commands.TryGetValue(lookup.ToString(), out var result)
                ? Result.Ok((result, splitIndex))
-               : defaultForCommands(lookup);
+               : defaultParserForCommands(lookup);
       } else
-        return nonCommands.TryLookup(consume) is { } result ? result : @default(consume);
+        return TryLookupNonCommand(chars);
+    }
+    Result<(TValue Result, int SplitIndex)> TryLookupNonCommand(ReadOnlySpan<char> chars) {
+      string? commandFound = null; // TODO:short-circuit when found
+      foreach (string command in nonCommands.Keys) {
+        if (chars.StartsWith(command.AsSpan(), StringComparison.InvariantCulture)) {
+          commandFound = command; }
+      }
+      return commandFound == null ? @defaultParser(chars) : Result.Ok((nonCommands[commandFound],commandFound.Length));
     }
   }
 
-  //https://stackoverflow.com/questions/255341/getting-key-of-value-of-a-generic-dictionary/255638#255638
+  // Taken from https://stackoverflow.com/questions/255341/getting-key-of-value-of-a-generic-dictionary/255638#255638
+  /// <summary>
+  /// Represents a many to one relationship between TFirsts and TSeconds, allowing fast lookup of the first TFirst corresponding to any TSecond.
+  /// </summary>
+#pragma warning disable CA1710 // Identifiers should have correct suffix
+#pragma warning disable CA1010 // Collections should implement generic interface
   public class BiDictionary<TFirst, TSecond>
-    : ProxyAdder<TFirst, TSecond>, IDictionary<TFirst, TSecond>, IReadOnlyDictionary<TFirst, TSecond> {
-    public BiDictionary(Action<TFirst, TSecond>? added = null) : base(added) =>
+#pragma warning restore CA1010 // Collections should implement generic interface
+#pragma warning restore CA1710 // Identifiers should have correct suffix
+    : ProxyAdder<TFirst, TSecond>
+    where TFirst: IEquatable<TFirst> {
+    public BiDictionary(Action<TFirst, TSecond>? extraCommandToPerformWhenAdding = null) : base(extraCommandToPerformWhenAdding) =>
       Added += (first, second) => {
         switch (firstToSecond.ContainsKey(first), secondToFirst.ContainsKey(second)) {
-          case (true, true):
-            firstToSecond.Add(first, second); // Throw: key and value both exist
-            break;
-          case (true, false):
-            secondToFirst.Add(second, first);
-            break;
+          case (true, _):
+            throw new Exception("Key already exists in BiDictionary.");
```

Comment:
```  
/// Represents a many to one relationship between TFirsts and TSeconds,
```
Behavior in the PR now matches this definition, which corresponds to the purpose of this type (a map from aliases to objects in which one alias is primary for each object mapped to). It's a good definiton because a more symmetric implementation (many to many, with one unidirectional link preferred in each case) is more complex and would require the type and methods to be redocumented and reimplemented.

Reply:
I think it would be better to rename the `BiDictionary` type to something that fits new definition.

Reply:
I believe it used to be called AliasMap

Reply:
There used to be both `AliasDictionary` and `BiDictionary`. Name name could be `AliasBiDictionary`.